astropy/coordinates/representation.py

```diff
@@ -421,39 +421,56 @@ class CartesianRepresentation(BaseRepresentation):
 
     Parameters
     ----------
-    x, y, z : `~astropy.units.Quantity`
-        The x, y, and z coordinates of the point(s). If ``x``, ``y``, and
-        ``z`` have different shapes, they should be broadcastable.
-
+    x, y, z : `~astropy.units.Quantity` or array
+        The x, y, and z coordinates of the point(s). If ``x``, ``y``, and ``z``
+        have different shapes, they should be broadcastable. If not quantity,
+        ``unit`` should be set.  If only ``x`` is given, it is assumed that it
+        contains an array with the 3 coordinates are stored along ``axis``.
+    unit : `~astropy.unit.Unit` or str
+        If given, the coordinates will be converted to this unit (or taken to
+        be in this unit if not given.
+    axis : int, optional
+        The axis along which the coordinates are stored for the case an array
+        of vectors is passed in (default: 0).
     copy : bool, optional
-        If True arrays will be copied rather than referenced.
+        If True (default), arrays will be copied rather than referenced.
     """
 
     attr_classes = OrderedDict([('x', u.Quantity),
                                 ('y', u.Quantity),
                                 ('z', u.Quantity)])
 
-    def __init__(self, x, y=None, z=None, copy=True):
+    def __init__(self, x, y=None, z=None, unit=None, axis=0, copy=True):
+
+        if unit is None and not hasattr(x, 'unit'):
+            raise TypeError('x should have a unit unless an explicit unit '
+                            'is passed in.')
 
         if y is None and z is None:
+            if axis != 0:
+                x = np.rollaxis(x, axis, 0)
             x, y, z = x
         elif (y is None and z is not None) or (y is not None and z is None):
-            raise ValueError("x, y, and z are required to instantiate CartesianRepresentation")
-
-        if not isinstance(x, self.attr_classes['x']):
-            raise TypeError('x should be a {0}'.format(self.attr_classes['x'].__name__))
-
-        if not isinstance(y, self.attr_classes['x']):
-            raise TypeError('y should be a {0}'.format(self.attr_classes['y'].__name__))
-
-        if not isinstance(z, self.attr_classes['x']):
-            raise TypeError('z should be a {0}'.format(self.attr_classes['z'].__name__))
-
-        x = self.attr_classes['x'](x, copy=copy)
-        y = self.attr_classes['y'](y, copy=copy)
-        z = self.attr_classes['z'](z, copy=copy)
+            raise ValueError("x, y, and z are required to instantiate {0}"
+                             .format(self.__class__.__name__))
```

Comment:
I think maybe there should be an exception now raised *also* in the case where `axis !=0` but `y` and `z` are provided.  Right now `axis` is just silently ignored in that case, right?  Catching that will help users understand what the point of `axis` is.